### PR TITLE
Refactor and add `num_nodes` and `verbose` options to `get_top_moves()`

### PR DESCRIPTION
**Pull request by knutole**
This PR combines and replaces https://github.com/py-stockfish/stockfish/pull/14 and https://github.com/py-stockfish/stockfish/pull/15 , with major refactor of `get_top_moves()` function. (It was much easier to combine the PR's, for obvious reasons.) I have left in quite a few comments in the refactor, to make it easier to follow. I will remove most of them once you've had a chance to look it over. 

- Refactor `get_top_moves()`, remove redundant code
- Add support for 'num_nodes', see #14
- Add support for 'verbose', see #15
- Add tests for new arguments
- Rename 'N/s' to 'NodesPerSecond'
- Add `turn_perspective` global option, defaults to _Turn_ perspective. (Note: this means that `get_top_moves()` will change from returning White perspective to returning Turn perspective by default, which is not backwards compatible. See https://github.com/py-stockfish/stockfish/issues/18)
- Add get/set for `self._turn_perspective`
- Add get/set for `self._num_nodes`
- Add `self._depth` as `int` and add public getter `get_depth()`. Fixes https://github.com/py-stockfish/stockfish/issues/8
- Add type checks for `set_depth()`, `set_num_nodes()`, `set_turn_perspective()` 
- Add more tests for instantiating `Stockfish` class
- Add test for `_set_option()`. Remove redundant default `update_parameters_attribute` arguments.
- Add convenience method `_pick()` to avoid doing `current_line[current_line.index("cp") + 1]`

Todo: Add `turn_perspective` option for rest of codebase, see https://github.com/py-stockfish/stockfish/issues/18

### CODE REVIEW

**Comment by kieferro** [comment]
Wow, that's quite a PR. First of all, a big thank you for the work and the interest that you put in this project!
I am currently not able to test and review your code, but it looks like you're still at work anyway. So just ping us or request if you need a review.

**Comment by knutole** [comment]
Thank you 🙏 This PR is ready for review 👌

**Comment by knutole** [comment]
It's not possible for me to request review on this PR, by the way, not sure if there's a Github setting for that. @kieferro @johndoknjas

File: stockfish/models.py [hunk 1]
```diff
@@ -519,75 +576,167 @@ def get_evaluation(self) -> dict:
...
+        # and the output is stored in the list 'lines'
+
         top_moves: List[dict] = []
-        multiplier = 1 if ("w" in self.get_fen_position()) else -1
-        for current_line in reversed(lines):
-            if current_line[0] == "bestmove":
-                if current_line[1] == "(none)":
+
+        # set perspective of evaluations. if get_turn_perspective() is True, or white to move,
+        # use Stockfish' values, otherwise invert values.
```
Comment (kieferro) [inline stockfish/models.py:649]:
```suggestion
        # use Stockfish's values, otherwise invert values.
```

File: stockfish/models.py [hunk 2]
```diff
@@ -519,75 +576,167 @@ def get_evaluation(self) -> dict:
...
+        if old_multipv != self._parameters["MultiPV"]:
+            self._set_option("MultiPV", old_multipv)
+
+        # reset self._num_nodes to global value
+        if old_num_nodes != self._num_nodes:
+            self._num_nodes = old_num_nodes
+
         return top_moves
 
+    def _pick(self, line: list = [], value: str = "", index: int = 1) -> str:
```
Comment (kieferro) [inline stockfish/models.py:737]:
```suggestion
    def _pick(self, line: list, value: str = "", index: int = 1) -> str:
```
Does it make sense to call the function without this parameter?

**Review by kieferro — CHANGES_REQUESTED** [review]
Okay, I am back and I looked at the changes. It looks very good overall!
We should, as you also wrote in your comment, implement the use of `turn_perspective` more consistently, but that's probably beyond the scope of this PR. This is definitely a good first step and we'll continue it from there.

**Comment by kieferro** [comment]
> It's not possible for me to request review on this PR, by the way, not sure if there's a Github setting for that. @kieferro @johndoknjas

Ah okay, I see. But [it seems that nothing can be done](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) about that as long as you don't have write-access. But feel free to just use the draft PR feature to indicate when the PR is ready for review.

File: stockfish/models.py [hunk 3]
```diff
@@ -327,6 +336,54 @@ def set_elo_rating(self, elo_rating: int = 1350) -> None:
             {"UCI_LimitStrength": "true", "UCI_Elo": elo_rating}
         )
 
+    def set_depth(self, depth: int = 15) -> None:
+        """Sets current depth of Stockfish engine.
+
+        Args:
+            depth: Depth as integer higher than 1
```
Comment (knutole) [inline stockfish/models.py:343]:
```suggestion
depth: Depth as integer 1 or higher
```

File: README.md [hunk 4]
```diff
@@ -135,16 +135,47 @@ True
...
+#   "Mate":"None",
+#   "Nodes":"2767506",
+#   "NodesPerSecond":"526442",
+#   "Time":"5257",
+#   "SelectiveDepth":"31",
+#   "MultiPVLine":"1",
+#   "WDL":"0 0 1000"
+# }]
+```
+Optional parameter `num_nodes` specifies the number of nodes to search. If num_nodes is 0, then the engine will search until it configured depth is reached.
```
Comment (kieferro) [inline README.md:163]:
```suggestion
Optional parameter `num_nodes` specifies the number of nodes to search. If num_nodes is 0, then the engine will search until it's configured depth is reached.
```

File: README.md [hunk 5]
```diff
@@ -135,16 +135,47 @@ True
...
+
+### Set perspective of the evaluation
```
Comment (kieferro) [inline README.md:165]: Maybe we should specify here that so far this setting only affects `get_top_moves`.

File: tests/stockfish/test_models.py [hunk 6]
```diff
@@ -448,9 +478,48 @@ def test_get_evaluation_stalemate(self, stockfish):
...
+        assert stockfish.get_depth() == 20
+        assert stockfish._depth == 20
+
+    def test_set_num_nodes(self, stockfish):
+        stockfish.set_num_nodes(100)
+        assert stockfish._num_nodes == 100
+        stockfish.set_num_nodes()
+        assert stockfish._num_nodes == 1000000
+
+    def test_set_num_nodes_raises_type_error(self, stockfish):
```
Comment (kieferro) [inline tests/stockfish/test_models.py:508]: Here I would also prefer to use `@pytest.mark.parametrize`. I must have overlooked this test last time.

**Review by kieferro — APPROVED** [review]

**Review by johndoknjas — APPROVED** [review]
The changes look good, and the tests pass for me - nice job! Thanks for your contribution.

**Comment by johndoknjas** [comment]
Looks like we're all satisfied with the PR, so I've merged it into master.

**Comment by knutole** [comment]
Great, thanks 🙏👍

**Comment by kieferro** [comment]
> Looks like we're all satisfied with the PR, so I've merged it into master.

Thanks :+1:

Congratulations @knutole. Could you close your two PRs in the old repo (https://github.com/zhelyabuzhsky/stockfish/pull/115, https://github.com/zhelyabuzhsky/stockfish/pull/116)? This would give us a better overview.

**Pull request merged**